Describe all cypher query options (#678)
neo4j/neo4j-documentation: cypher/cypher-docs/src/docs/dev/query-tuning.asciidoc

|===
| Query option | Description | Default
| `3.5` | This will force the query to use Neo4j Cypher 3.5. |
| `4.1` | This will force the query to use Neo4j Cypher 4.1. |

> | `4.2` | This will force the query to use Neo4j Cypher 4.2. As this is the default version, it is not necessary to use this option explicitly. | X

Reviewer: What happened to the 4.0 version? Is it no longer available?

Reviewer 2: It is not available in 4.2 any longer, that is right.

| `connectComponentsPlanner=idp`
| Use the cost based IDP search algorithm when combining sub-plans.
  Note that using this option may significantly increase the planning time of the query. |
|===

[[cypher-update-strategy]]
=== Cypher update strategy
This option affects eagerness of updating queries
[options="header"]
|===
| Query option | Description | Default
| `updateStrategy=default` | Update queries are executed eagerly when needed. | X
| `updateStrategy=eager`   | Update queries are always executed eagerly. |
|===

[[cypher-expression-engine]]
=== Cypher expression engine
This option affects how the runtime evaluates expressions. The possible values are:
[options="header"]
|===
| Query option | Description | Default
| `expressionEngine=default`     | Compile expressions and use the compiled expression engine when needed. | X
| `expressionEngine=interpreted` | Always use the interpreted expression engine. |
| `expressionEngine=compiled`    | Always compile expressions and use the compiled expression engine. Cannot be used together with `runtime=interpreted` |
|===

[[cypher-operator-engine]]
=== Cypher operator engine

> This query option affects whether the pipelined runtime attempts to fuse groups of operators into pipelines. The possible values are:
> [options="header"]
> |===
> | Query option | Description | Default
> | `operatorEngine=default`     | Attempt to fuse operators when applicable. | X
> | `operatorEngine=interpreted` | Never attempt to fuse operators. |
> | `operatorEngine=compiled`    | Always attempt to fuse operators. Cannot be used together with `runtime=interpreted` or `runtime=slotted` |

Reviewer 2: "fuse" might sound misleading when the option name is "compiled".

Maybe this is enough?:
```suggestion
This query option affects whether the pipelined runtime attempts to generate compiled code for groups of operators. The possible values are:
[options="header"]
|===
| Query option | Description | Default
| `operatorEngine=default`     | Attempt to generate compiled operators when applicable. | X
| `operatorEngine=interpreted` | Never attempt to generate compiled operators. |
| `operatorEngine=compiled`    | Always attempt to generate compiled operators. Cannot be used together with `runtime=interpreted` or `runtime=slotted` |
```

This query option affects how the pipelined runtime behaves for operators it does not directly support. The available options are:
[options="header"]
|===
| Query option | Description | Default
| `interpretedPipesFallback=default`
| Equivalent to `interpretedPipesFallback=whitelisted_plans_only`
| X
| `interpretedPipesFallback=disabled`
| If the plan contains any operators not supported by the pipelined runtime then another runtime is chosen to execute the entire plan.
  Cannot be used together with `runtime=interpreted` or `runtime=slotted`
|
| `interpretedPipesFallback=whitelisted_plans_only`
| Parts of the execution plan may be executed on another runtime.
  Only certain operators are allowed to execute on another runtime.
  Cannot be used together with `runtime=interpreted` or `runtime=slotted`
|
| `interpretedPipesFallback=all`
| Parts of the execution plan may be executed on another runtime.
  Any operator is allowed to execute on another runtime.
  This setting is experimental and using it is discouraged.
  Queries with this option set might produce incorrect results or fail.
  Cannot be used together with `runtime=interpreted` or `runtime=slotted`
|

> |===

Reviewer 2: Having each cell on a new line reads a bit weird when looking at the bare asciidoc. I would prefer having each table row correspond to a line of source code.

Author: Well, the lines get really long.. The same thing is done in the `cypher-runtime`section above

Reviewer: I can clean up the asciidoc, but this will render as a single row in the table, as per the other tables on this page. 

I'm just reviewing now, and will push some editorial changes.
